### PR TITLE
Add prevent-type permissions for /itemname

### PR DESCRIPTION
**Pull request by JRoy**
Adds `essentials.itemname.prevent-type.<material_name>` permissions.

Reference:

| Material \ Wildcard | unset | false | true  |
|---------------------|-------|-------|-------|
| unset               | allow | allow | deny  |
| false               | allow | allow | allow |
| true                | deny  | deny  | deny  |

Closes #4044.

### CODE REVIEW

**Review by pop4959 — DISMISSED** [review]
Should use an allow-type permission instead of a prevent-type, otherwise we'll have the same mess as we do with hat permissions currently.

The issue is that implementing a negative permission like prevent-type will revoke access (which is a behavior change in Essentials) for servers that utilize wildcards. On the contrary, allow-type could just be set to false for specific materials, and servers using wildcards for this will still grant themselves access to all items by default.

**Comment by JRoy** [comment]
> Should use an allow-type permission instead of a prevent-type, otherwise we'll have the same mess as we do with hat permissions currently.
> 
> The issue is that implementing a negative permission like prevent-type will revoke access (which is a behavior change in Essentials) for servers that utilize wildcards. On the contrary, allow-type could just be set to false for specific materials, and servers using wildcards for this will still grant themselves access to all items by default.

Fixed

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commanditemname.java [hunk 1]
```diff
@@ -21,7 +20,14 @@ public Commanditemname() {
...
         final ItemStack item = user.getBase().getItemInHand();
         if (item.getType() == Material.AIR) {
-            user.sendMessage(tl("itemnameInvalidItem", item.getType().toString().toLowerCase(Locale.ENGLISH).replace('_', ' ')));
+            user.sendMessage(tl("itemnameInvalidItem"));
+            return;
+        }
+
+        final Boolean material = user.isAuthorizedStrict(PERM_PREFIX + item.getType().name().toLowerCase());
+        final Boolean wildcard = user.isAuthorizedStrict(PERM_PREFIX + "*");
+        if (((wildcard != null && wildcard) && material == null) || ((wildcard != null && wildcard) && material) || ((wildcard == null || !wildcard) && (material != null && material))) {
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commanditemname.java]:
```suggestion
        if (((wildcard != null && wildcard) && material == null) || ((wildcard != null && wildcard) && (material != null && material)) || ((wildcard == null || !wildcard) && (material != null && material))) {
```

Same logic as `Commandhat`

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandhat.java [hunk 2]
```diff
@@ -33,7 +33,9 @@ protected void run(final Server server, final User user, final String commandLab
                 return;
             }
 
-            if (user.isAuthorized("essentials.hat.prevent-type." + hand.getType().name().toLowerCase())) {
+            final Boolean wildcard = user.isAuthorizedStrict(PERM_PREFIX + "*");
+            final Boolean material = user.isAuthorizedStrict(PERM_PREFIX + hand.getType().name().toLowerCase());
+            if (((wildcard != null && wildcard) && material == null) || ((wildcard != null && wildcard) && material) || ((wildcard == null || !wildcard) && (material != null && material))) {
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandhat.java]:
```suggestion
            if (((wildcard != null && wildcard) && material == null) || ((wildcard != null && wildcard) && (material != null && material)) || ((wildcard == null || !wildcard) && (material != null && material))) {
```

For future reference:

| Material \ Wildcard | unset | false | true  |
|---------------------|-------|-------|-------|
| unset               | allow | allow | deny  |
| false               | allow | allow | allow |
| true                | deny  | deny  | deny  |

**Comment by kennytv** [comment]
Would look JRoy times better if you had a tristate enum à la
```java
if (wildcard == Thing.TRUE && (material == Thing.UNSET || material == Thing.TRUE))
                || (material == Thing.TRUE && (wildcard == Thing.UNSET ||  wildcard == Thing.FALSE));
```
and you pulled the || on material and wilcard in each line into an extra var as well

File: Essentials/src/main/java/com/earth2me/essentials/User.java [hunk 3]
```diff
@@ -701,7 +727,7 @@ public void updateActivityOnChat(final boolean broadcast) {
     }
 
     public void checkActivity() {
-        // Graceful time before the first afk check call. 
+        // Graceful time before the first afk check call.
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/User.java]:
```suggestion
        // Graceful time before the first afk check call. 
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commanditemname.java [hunk 4]
```diff
@@ -21,7 +21,14 @@ public Commanditemname() {
...
         if (item.getType() == Material.AIR) {
-            user.sendMessage(tl("itemnameInvalidItem", item.getType().toString().toLowerCase(Locale.ENGLISH).replace('_', ' ')));
+            user.sendMessage(tl("itemnameInvalidItem"));
+            return;
+        }
+
+        final TriState wildcard = user.isAuthorizedExact(PERM_PREFIX + "*");
+        final TriState material = user.isAuthorizedExact(PERM_PREFIX + item.getType().name().toLowerCase());
+        if ((wildcard == TriState.TRUE && material != TriState.FALSE) || ((wildcard != TriState.TRUE) && material == TriState.TRUE)) {
+            user.sendMessage(tl("itemnameInvalidItem"));
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commanditemname.java:31]: Maybe a no permission message here?

**Review by mdcfe — APPROVED** [review]

**Pull request merged**